### ISSUE TITLE
`std::ptr::drop_in_place` docs suggest overwriting value with `write`, which can lead to unsoundness in common cases

### ISSUE THREAD

**Issue by awused** [rust-lang/rust#127939]
### Location

[std](https://doc.rust-lang.org/std/index.html)::[ptr](https://doc.rust-lang.org/std/ptr/index.html)::[drop_in_place](https://doc.rust-lang.org/std/ptr/fn.drop_in_place.html#)

>[write()](https://doc.rust-lang.org/std/ptr/fn.write.html) can be used to overwrite data without causing it to be dropped

### Summary

I was looking for a way to drop a `!Unpin` future in place in https://github.com/StoicDeveloper/mapped_futures/issues/4 and thought `drop_in_place` was the obvious choice, and after reading the safety docs, which specifically mention write(), I came up with this inadvertently unsound code:

```rust
// Cell is an UnsafeCell<Option<impl !Unpin>>
unsafe {
    cell.get().drop_in_place();
    cell.write(None);
}
```

Admittedly, this was my mistake for not reading the example closely, which makes it obvious, but I was probably reading what I wanted to read at the time because `drop_in_place` was the "obvious" correct choice for dropping a Future in place. But I think if the example wasn't there, it would not be clear to many (most?) readers that this code was unsound from the `Safety` section alone.

While I think the unsoundness of my code is implied by 
`While drop_in_place is executing, the only way to access parts of to_drop...` and `using the pointed-to value after calling...` it's unclear that it includes Drop handlers for containers that may attempt to drop value again during unwinding. I think it would help if the docs specifically mentioned panicking and unwinding.

It would also help if it suggested alternate ways to drop and replace a value in place. I would not have guessed `*cell.get() = None` is the best option it is (used internally in futures-rs). But it's not obvious that that was an option since it looks like it could move the value, where `cell.get().drop_in_place()` _looks_ more correct because it is explicit about dropping without moving.

**Comment by workingjubilee** [comment]
@awused You mean `cell.get().write`, right?

**Comment by awused** [comment]
> @awused You mean `cell.get().write`, right?

Yeah, my copy-paste mistake.

**Comment by workingjubilee** [comment]
My understanding of some of the base facts:
- `Pin<P>` wraps a "pointer-like" type, `P`, which itself is some kind of `Ptr<T>`.
- `Pin<Ptr<T>>` works to prevent `Ptr<T>` from "moving" `T` if `T: !Unpin`.
- `T: !Unpin` does not mean that it must not be moved! ...It means *pointers* should not move it, and then, only if they are within `Pin`.
- Code that owns the true value, instead of having an indirection, may still "simply" do something that executes the destructor. So you should still "simply" do that when you want to execute the destructor.

@Darksonn Is that correct?

My understanding of this current issue:
- It is easy to, reading the `drop_in_place` documentation, think that calling `ptr::drop_in_place` and then `ptr::write`, with the same pointer, in that order(!), is acceptable.
- It's probably incorrect, and even if it is sound in _some_ case[^1], doing it in that order is _probably_ unsound when combined with the sort of situation that prompts people to carefully read the docs to `drop_in_place`, such as unsafe async code!
- Thus the docs should be clearer that using `ptr::drop_in_place` and `ptr::write` are *alternatives* to each other, and you should probably do only one or the other. I guess it's sound to `ptr::write` and then `ptr::drop_in_place`, in that order, so that you run the side effect of a drop? But the reverse is probably never what you wanted.

And this all combines with panics which make things _quite_ exciting.

[^1]: I'm not sure if it's always incorrect, and the answer is probably partly dependent on undecided opsem questions?

**Comment by Darksonn** [comment]
> It is easy to, reading the `drop_in_place` documentation, think that calling `ptr::drop_in_place` and then `ptr::write`, with the same pointer, in that order(!), is acceptable.

The tricky thing here is that `ptr::drop_in_place` followed by `ptr::write` is actually the right operations in the right order. It's just that you need to execute the `ptr::write` operation _even if `ptr::drop_in_place` panics_. That doesn't happen when you write this:
```Rust
unsafe {
    cell.get().drop_in_place();
    cell.get().write(None); // This line doesn't run if drop_in_place panics
}
```
whereas `*cell.get() = None` runs the destructor and then writes `None` even if the destructor panics.

Note that this operation is equivalent to using [`Pin::set`](https://doc.rust-lang.org/stable/std/pin/struct.Pin.html#method.set).

> `T: !Unpin` does not mean that it must not be moved! ...It means pointers should not move it, and then, only if they are within `Pin`.
>
> @Darksonn Is that correct?

Not quite. If a value is pinned, it must not be moved by anyone under any circumstances. This is true even if there are no longer any pinned pointers left!

Though, note that `T: !Unpin` is not equivalent to "this value is pinned". A `T: !Unpin` value is not pinned until you create a `Pin<Ptr<T>>` pointing at the value. Before such a pinned pointer exists, the value may be moved. But once a pinned pointer has been created, the value must not be moved by anyone until its destructor runs. This applies even if the `Pin<Ptr<T>>` does not exist anymore — pinning cannot be undone.

To summarize, the "no moving" requirement applies from the creation of the first `Pin<Ptr<T>>` value, and until `drop` runs on that memory location.

**Comment by workingjubilee** [comment]
Exciting, we get to learn things today!

**Comment by awused** [comment]
I think most people who end up on the docs for `drop_in_place` have the right idea, but are very likely in the wrong place. The safe use cases for `drop_in_place` outside of a `Drop` implementation are probably very rare.

For the alternatives `ManuallyDrop::drop()`, assigning to `(* mut)`, or sometimes `Box/Rc/Arc::from_raw` depending on how it was created, or it's not always obvious that those will drop the value in place without moving them. If that were obvious, the reader would probably not be in the docs for `drop_in_place`. It's not even immediately obvious that writing to `(*mut)` will run the drop code; it is present in the ptr docs, but those don't mention it being in place and dealing with `?Unpin` is rare enough to want explicit guarantees. For a person who is uncertain (myself, in this case), the lack of explicit guarantees led me to the explicit but wrong `drop_in_place`.
